base/abn_metrics.go

```diff
@@ -0,0 +1,174 @@
+package base
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	abnapp "github.com/iter8-tools/iter8/abn/application"
+	pb "github.com/iter8-tools/iter8/abn/grpc"
+	log "github.com/iter8-tools/iter8/base/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+const (
+	defaultEndpoint = "iter8-abn:50051"
+)
+
+// abnClientInterface is interface for calling gRPC services
+type abnClientInterface interface {
+	callGetApplicationJSON(appName string) (string, error)
+}
+
+// defaultABNClient is default implementation of interface that calls the service
+type defaultABNClient struct {
+	endpoint string
+}
+
+func (wc *defaultABNClient) callGetApplicationJSON(appName string) (string, error) {
+	// setup client
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+	conn, err := grpc.Dial(wc.endpoint, opts...)
+	if err != nil {
+		return "", err
+	}
+	c := pb.NewABNClient(conn)
+
+	// get application
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	s, err := c.GetApplicationData(
+		ctx,
+		&pb.ApplicationRequest{
+			Application: appName,
+		},
+	)
+	if err != nil {
+		return "", err
+	}
+	applicationJSON := s.GetApplicationJson()
+
+	return applicationJSON, nil
+}
+
+const (
+	// CollectABNMetrics is the name of the task to read A/B(/n) metric values
+	CollectABNMetrics = "abnmetrics"
+	// abnMetricProvider is prefix for abn metrics
+	abnMetricProvider = "abn"
+)
+
+// ABNMetricsInputs is the inputs for for the abnmetrics task
+type ABNMetricsInputs struct {
+	Endpoint *string `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+	// Application is name of application to evaluate
+	Application string `json:"application" yaml:"application"`
+}
+
+// collectABNMetricsTask is task definition for abnmetrics task
+type collectABNMetricsTask struct {
+	TaskMeta
+	With   ABNMetricsInputs `json:"with" yaml:"with"`
+	client abnClientInterface
+}
+
+// initializeDefaults sets default values for the task
+func (t *collectABNMetricsTask) initializeDefaults() {
+	if t.client == nil {
+		ep := defaultEndpoint
```

Comment:
Could we please 

1. Eliminate `defaultEndpoint`
2. Make `t.With.Endpoint` required (i.e., add it to validateInputs, and make sure it exists)
3. Set:
```
t.client = &defaultABNClient{
  endpoint: t.With.Endpoint
}
```

Basically, we do not want to deal with this setting of this end point within this task at all (but simply require that whoever invokes the task pass it along).

Especially, since we said Helm chart would set a default value (we don't want two defaults, or two different places where they are managed).

Reply:
done.